openmls/src/group/core_group/create_commit.rs

```diff
@@ -38,9 +38,41 @@ impl CoreGroup {
 
         // If this is an external commit, we don't have an `own_leaf_index` set
         // yet. Instead, we use the index in which we will be put in course of
-        // this commit.
+        // this commit. Our index is determined as if we'd be added through an
+        // Add proposal. However, since this might be the "resync" flavour of an
+        // external commit, it could be that we're first removing our past self
+        // from the group, in which case, we can't just take the next free leaf
+        // in the existing tree. Note, that we have to determine the index here
+        // (before we actually add our own leaf), because it's needed in the
+        // process of proposal filtering and application.
         let (own_leaf_index, sender_type) = match params.commit_type() {
-            CommitType::External => (self.treesync().free_leaf_index()?, SenderType::NewMember),
+            CommitType::External => {
+                // If this is a "resync" external commit, it should contain a
+                // `remove` proposal with the index of our previous self in the
+                // group.
+                let free_leaf_index = self.treesync().free_leaf_index()?;
+                let remove_proposal_option = params
+                    .inline_proposals()
+                    .iter()
+                    .find(|proposal| proposal.is_type(ProposalType::Remove));
+                // The external commit is processed like an add, so our new leaf
+                // index is the same as if we'd process an add after the remove
+                // proposal.
+                let leaf_index = if let Some(remove_proposal) = remove_proposal_option {
+                    let removed_index = remove_proposal
+                        .as_remove()
+                        .ok_or(CoreGroupError::LibraryError)?
+                        .removed();
+                    if removed_index < free_leaf_index {
```

Comment:
I'm not sure I understand this logic... In the case of a "resync", are we moving ourselves in the tree to the left if possible?

Reply:
We're not really moving ourselves. We're removing our old self and add ourselves into the group like a new member. They leaf index does not stay consistent.